pkg/controller/apimanager/upgrade_26_to_27.go

```diff
@@ -0,0 +1,383 @@
+	existingDeploymentConfig := &appsv1.DeploymentConfig{}
+	err := u.client.Get(context.TODO(), types.NamespacedName{Name: "system-sidekiq", Namespace: u.cr.Namespace}, existingDeploymentConfig)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	envVarToRemove := "AMP_RELEASE"
+
+	removed := u.removeEnvVarFromDCContainers(existingDeploymentConfig, envVarToRemove)
+
+	if removed {
+		err := u.client.Update(context.TODO(), existingDeploymentConfig)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (u *Upgrade26_to_27) removeEnvVarFromDCContainers(deploymentConfig *appsv1.DeploymentConfig, envVarName string) bool {
+	removed := false
+
+	for idx := range deploymentConfig.Spec.Template.Spec.Containers {
+		container := &deploymentConfig.Spec.Template.Spec.Containers[idx]
+		containerEnv := container.Env
+		for envIdx, val := range containerEnv {
+			if val.Name == envVarName {
+				removed = true
+				container.Env = append(container.Env[:envIdx], container.Env[envIdx+1:]...)
+			}
+		}
+	}
+
+	return removed
+}
+
+func (u *Upgrade26_to_27) upgradeImageStream(imageStream *imagev1.ImageStream, newImageURL string) (bool, error) {
+	changed := false
+	tagToRemove := "2.6"
+	desiredTag := product.ThreescaleRelease
+
+	// TODO there are three options to approach this:
+	// 1 - Replace the name of the the old version-tag with the new one. This is
+	//     the approach that has been implemented here
+	// 2 - Delete the version-tag and create a new version tag
+	// 3 - Add the new version tag
+	imageStreamSpec := &imageStream.Spec
+	for idx := range imageStreamSpec.Tags {
+		tag := &imageStreamSpec.Tags[idx]
+		if tag.Name == "latest" && tag.From.Name == tagToRemove {
+			tag.From.Name = desiredTag
+			changed = true
+		}
+
+		if tag.Name == tagToRemove {
+			tag.Name = desiredTag
+			changed = true
+		}
+	}
+
+	if changed {
+		err := u.client.Update(context.TODO(), imageStream)
+		if err != nil {
+			return changed, err
+		}
+	}
+
+	return changed, nil
+}
+
+func (u *Upgrade26_to_27) upgradeAPIcastImageStream() (bool, error) {
+	// if u.cr.Spec.Backend != nil && u.cr.Spec.Backend.Image != nil {
+	// 	// TODO When the image field is overwritten should we apply the upgrade procedure
+	// 	// for the image? In case we don't apply it, how should we behave? by returning an error
+	// 	// saying that upgrade procedure failed and to remove those fields ?? or by simply
+	// 	// skipping that upgrade part? For the moment the second approach has been taken
+	// 	// for simplicity in this method but probably this will have to be changed
+	// 	return false, nil
+	// }
+
+	apicastImageStream := &imagev1.ImageStream{}
+	err := u.client.Get(context.TODO(), types.NamespacedName{Name: "amp-apicast", Namespace: u.cr.Namespace}, apicastImageStream)
+	if err != nil {
+		return false, err
+	}
+
+	newImageURL := product.CurrentImageProvider().GetApicastImage()
```

Comment:
Yeah , this is the biggest issue: new image URL is hardcoded or they can be read from CR?

I think that we should read from CR and desired state enforces some images, let it be.

If hardcoded, upgrade will update images to hardcoded ones, but inmediate reconciliacion iteration will update then to the desired state.

We should make very clear that upgrading procedure could potentially fail if images are overridden in CR. 

WDYT??

Reply:
I agree with what you say.
I've updated the PR and now in an image has been overriden at the CR level, during the ImageStreams upgrade procedure, it does not apply the corresponding imagestream upgrade logging a warning about it and continuing the upgrade process.
I agree that we should make very clear that the upgrade procedure could fail if images are overriden in the CR and that the users should never set them unless they are in some devtest scenario and knowing that they won't be able to correctly upgrade in that case.

As we talked before, I'm a bit concerned about how we will manage the upgrade procedure testing when doing QE as the procedure until now has always required overriding the images and this would make the upgrade not testable, but lets do it this way for the moment as our intention is that users should not override when upgrading, and lets keep talking with QE on how are we going to manage that specific scenario.